edb/pgsql/types.py

```diff
@@ -440,18 +399,82 @@ def get_pointer_storage_info(
         schema, material_ptrcls = pointer.material_type(schema)
     if material_ptrcls is not None:
         pointer = material_ptrcls
-    return _PointerStorageInfo(
-        schema, pointer, source=source, resolve_type=resolve_type,
-        link_bias=link_bias)
 
+    if source is None:
+        source = pointer.get_source(schema)
+
+    is_lprop = pointer.is_link_property(schema)
 
-class PointerStorageInfo(NamedTuple):
+    if resolve_type and schema is None:
+        msg = 'PointerStorageInfo needs a schema to resolve column_type'
+        raise ValueError(msg)
+
+    if is_lprop and pointer.issubclass(
+        schema, schema.get('std::target', type=s_obj.SubclassableObject)
+    ):
+        # Normalize link@target to link
+        assert isinstance(source, s_pointers.Pointer)
+        pointer = source
+        is_lprop = False
+
+    if isinstance(pointer, irast.TupleIndirectionLink):
+        table = None
+        table_type = 'ObjectType'
+        col_name = pointer.get_shortname(schema).name
+    elif is_lprop:
+        table = common.get_backend_name(schema, source, catenate=False)
+        table_type = 'link'
+        if pointer.get_shortname(schema).name == 'source':
+            col_name = 'source'
+        else:
+            col_name = str(pointer.id)
+    else:
+        if isinstance(source, s_scalars.ScalarType):
+            # This is a pseudo-link on an scalar (__type__)
+            table = None
+            table_type = 'ObjectType'
+            col_name = None
+        elif _pointer_storable_in_source(schema, pointer) and not link_bias:
+            table, table_type, col_name = _source_table_info(schema, pointer)
+        elif _pointer_storable_in_pointer(schema, pointer):
+            table, table_type, col_name = _pointer_table_info(schema, pointer)
+        else:
+            return None  # type: ignore
+
+    if resolve_type:
+        column_type = _resolve_type(schema, pointer)
+    else:
+        column_type = None
+
+    return PointerStorageInfo(
+        table_name=table,
+        table_type=table_type,
+        column_name=col_name,  # type: ignore
+        column_type=column_type,  # type: ignore
+    )
+
+
+@dataclasses.dataclass(kw_only=True, repr=False, eq=False, slots=True)
+class PointerStorageInfo:
 
     table_name: Optional[Tuple[str, str]]
     table_type: str
     column_name: str
     column_type: Tuple[str, str]
 
+    def __repr__(self) -> str:
```

Comment:
didn't like the default dataclass repr?

Reply:
I've copied this from `_PointerStorageInfo`, thinking it's a part of some clever framework for `dump()`. Apparently not.